File: workspace/2023-03-12/factorial.erlang/factorial.erl

main(_) ->
    show_factorials(10).

show_factorials(N) ->
    show_factorials(0, N).

show_factorials(Min, Max)
    when
        Min =< Max ->
    show_factorial(Min),
    show_factorials(Min + 1, Max);

show_factorials(_, _) ->
    ok.

show_factorial(N) ->
    io:format("factorial(~w) = ~w~n", [N, factorial(N)]).

factorial(0) ->
    1;

factorial(N) ->
    N * factorial(N - 1).

File: workspace/2023-04-01/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fn n -- fn )
    dup 0 = if
        drop        ( fn n -- fn )
    else
        dup         ( fn n -- fn n n )
        rot *       ( fn n n -- n n*fn )
        swap        ( n n*fn -- n*fn n )
        1 -         ( n*fn n -- n*fn n-1 )
        _factorial  ( n*fn n-1 -- new_fn )
    endif
;

: factorial ( n -- fn )
    1 swap      ( n -- 1 n )
    _factorial  ( 1 n -- fn )
;

: _show_factorials recursive ( max min -- )
    over over < if
        drop drop           ( max min -- )
    else
        dup factorial .
        1 +                 ( max min -- max min+1 )
        _show_factorials    ( max min+1 -- )
    endif
;

: show_factorials ( n -- )
    1                   ( n -- n 1 )
    _show_factorials    ( n 1 -- )
;

